[libunwind] Support rtld-c18n as the runtime linker.

dstolfa opened the pull request:
This is an initial take on a possible implementation to support the c18n runtime linker ABI in libunwind. The last commit message should describe the approach. I will note that we could consider putting the `CompartmentInfo` data in the unwind cursor, but my worry is that whatever is calling it will have access to that data, which would leak other compartments' restricted stack capabilities. Perhaps there's a way to reliably limit access to that part of the capability that I'm not aware of? A temporary "memory leak" sounded like a better idea otherwise (as described in the commit message), even though it makes me a bit uneasy.

This depends on https://github.com/CTSRD-CHERI/cheribsd/pull/2032 in order to actually work with the c18n runtime linker, but libunwind should continue to work as normal without that patch, it just won't be able to unwinding compartments. The two are bundled together in https://github.com/CTSRD-CHERI/cheribsd/pull/2003 for ease of building.

dstolfa commented:
Updated the review to address a comment from @jrtc27 on another PR noting that concurrent unwinding of multiple contexts and cursors from the same thread is a part of the current libunwind API. Moved the pointer to unwind information for compartments to the caller stack, but optionally (`_LIBUNWIND_SANDBOX_HARDENED`) seal the pointer so that the caller cannot dereference or modify it. Since the change was already ABI breaking for both the context and cursor, this is probably okay to do as well.

Note that we use the heap to allocate the actual hash table as it essentially quadruples the size of the unwind cursor on the stack otherwise (going from ~1K to 4K per cursor).

jrtc27 commented:
Doesn't this leak the tables? I don't think the destructor of UnwindCursor, and thus of CompartmentInfo, is normally called, given it's normally just a stack-allocated unw_cursor_t which is reinterpreted as an UnwindCursor.

dstolfa commented:
> Doesn't this leak the tables? I don't think the destructor of UnwindCursor, and thus of CompartmentInfo, is normally called, given it's normally just a stack-allocated unw_cursor_t which is reinterpreted as an UnwindCursor.

You're right, it does. I'd completely glossed over the fact that the destructor wouldn't be called. This makes the problem slightly harder because I don't see an easy way to use the heap here, which then makes me concerned about cases where we have lots of compartments that we need to unwind through.

Obviously for backtraces and exception handling we could clean up in the respective functions, but it doesn't address the wider problem of libunwind APIs. I'll think about it a bit more as I'd really like to avoid both introducing limitations on how many compartments we can unwind and introducing any new API to libunwind.

dpgao commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -403,9 +567,38 @@ int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pc_t pc,
+        uintcap_t csp = registers.getUnsealedECSP(sealer);
+	CompartmentInfo &CI = CompartmentInfo::sThisCompartmentInfo;
+        for (;;) {
+          if (isEndOfExecutiveStack(csp, CI)) {
```
Is it possible that we are at the end of the Executive stack and yet still have frames to unwind? For example, we are in RTLD and have not made any compartment transitions, so the Executive stack would contain no frame.

To handle situations like this, I guess we can do something like
```
while (!isEndOfExecutiveStack(...) && isTrampoline(...)) {
  returnAddress = ...
  csp = ...
}
```

dstolfa replied:
Is that a situation we would ever end up in with libunwind? For us to begin unwinding, someone would have to have called `unw_getcontext` and `unw_init_local`, presumably in one of the compartments.

dpgao commented on libunwind/src/Registers.hpp:
```diff
@@ -1983,9 +2034,25 @@ inline bool Registers_arm64::validCapabilityRegister(int regNum) const {
+  assert(sealer != (uintcap_t)-1 && "Sealer not initialized");
+  uintcap_t csp = _registers.__csp;
+#ifdef _LIBUNWIND_SANDBOX_HARDENED
+  if (__builtin_cheri_sealed_get(csp))
+    return __builtin_cheri_unseal(csp, sealer);
+  else
```
Shorter to write
```
if (__builtin_cheri_sealed_get(csp))
  csp = __builtin_cheri_unseal(csp, sealer);
```

dpgao commented on libunwind/src/Registers.hpp:
```diff
@@ -1851,6 +1852,43 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+             "Value must be sealed");
+#endif
+      uintcap_t unsealedValue = sealedValue;
+      if (__builtin_cheri_sealed_get(sealedValue))
+        unsealedValue = __builtin_cheri_unseal(sealedValue, sealer);
+      // If the tag gets cleared when we attempt to unseal our value, that means
+      // that we either have a capability that was sealed to begin with, and
+      // therefore we should just return it that way, or we have a sentry which
+      // we cannot unseal.
+      if (!__builtin_cheri_tag_get(unsealedValue) &&
+          __builtin_cheri_tag_get(sealedValue))
+        unsealedValue = sealedValue;
```
Might be clearer to write this show that the second `if` is really dependent on the first `if`
```
if (__builtin_cheri_tag_get(sealedValue) && __builtin_cheri_sealed_get(sealedValue)) {
    unsealedValue = __builtin_cheri_unseal(sealedValue, sealer);
    if (!__builtin_cheri_tag_get(unsealedValue))
        unsealedValue = sealedValue;
}
```

dpgao commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -297,25 +450,36 @@ int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pc_t pc,
-            // Leaf function keeps the return address in register and there is no
-            // explicit intructions how to restore it.
-            returnAddress = registers.getRegister(cieInfo.returnAddressRegister);
+          // Leaf function keeps the return address in register and there is no
+          // explicit intructions how to restore it.
+          returnAddress = registers.getRegister(cieInfo.returnAddressRegister);
```
Seems to be a diff caused by whitespace

dstolfa replied:
Yes, it was badly indented so `clang-format` ate a whitespace there I suppose. I guess I just left it because I also added `CHERI_DBG` right under it.

dpgao commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -703,6 +703,20 @@ Lnovec:
+#if defined(__CHERI_PURE_CAPABILITY__)
+DEFINE_LIBUNWIND_FUNCTION(__rtld_unw_setcontext)
+  mov    c16, c3
+  ldp    c3, c4,  [c4, #0x030]
```
The offset should be negative given that there was
```
add c4, c0, #0x210
```
before?

dstolfa replied:
I think so, yes.

dstolfa commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -941,6 +954,12 @@ class UnwindCursor : public AbstractUnwindCursor{
   virtual const char *getRegisterName(int num);
+#if defined(__CHERI_PURE_CAPABILITY__) &&                                      \
+    defined(_LIBUNWIND_SANDBOX_OTYPES) && defined(_LIBUNWIND_SANDBOX_HARDENED)
+  virtual void       unsealSP(uintcap_t sealer = -1);
+  virtual void       unsealFP(uintcap_t sealer = -1);
+  virtual void       unsealCalleeSavedRegisters(uintcap_t sealer = -1);
```
Yeah, they don't currently. They were probably used at one point, though I'm somewhat tempted to keep them as this is the value for the sealer that means "we are not running with c18n". I'm also happy to get rid of them if that's the overall preference.

dstolfa commented:
Address some of the above comments and import changes from the demo branch.

dstolfa commented:
Updated to support the changes in rtld.

bsdjhb commented on libunwind/src/Registers.hpp:
```diff
@@ -1882,6 +1928,9 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
     uintptr_t __sp;    // Stack pointer r31
     uintptr_t __pc;    // Program counter
+#ifdef __CHERI_PURE_CAPABILITY__
+    uintptr_t __csp; // Executive stack pointer.
```
Maybe call this `__ecsp` to reflect the comment?

bsdjhb commented on libunwind/src/Registers.hpp:
```diff
@@ -1924,6 +1973,8 @@ inline bool Registers_arm64::validRegister(int regNum) const {
 #ifdef __CHERI_PURE_CAPABILITY__
   if ((regNum >= UNW_ARM64_C0) && (regNum <= UNW_ARM64_C31))
     return true;
+  if (regNum == UNW_ARM64_ECSP)
```
Do we need to permit clients to fetch the value of ECSP?  If not, we should remove `UNW_ARM64_ECSP` and support for it via the public register interface.

dstolfa replied:
Deleted this, we don't really want clients fetching this in any circumstance.

bsdjhb commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -54,6 +55,15 @@ class DwarfInstructions {
   typedef typename CFI_Parser<A>::FDE_Info          FDE_Info;
   typedef typename CFI_Parser<A>::CIE_Info          CIE_Info;
 
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(_LIBUNWIND_SANDBOX_OTYPES)
```
Can we just use the libunwind native types here?  That is, just use `pint_t` instead of `uintptr_t` and probably use it instead of `uintcap_t` as well?  In particular, in purecap code, we don't need to use `uintcap_t` and can instead just use `uintptr_t` normally.  Typically `uintcap_t` is only used in hybrid code and all of this is purecap.

dstolfa replied:
As discussed, `pint_t` is now used everywhere except for `Registers.hpp` which can't include `AddressSpace.hpp`.

bsdjhb reviewed (commented):
I think it would be easier to review (and easier to re-read in the future for history purposes) if the optional bits were split out into follow-up commits.  That is, first a commit to add the base c18n unwinder, then a commit to add the first option (and include the cmake build glue as part of that commit), then a third commit to add the second option.

dstolfa commented:
Rebased on `dev`, merged https://github.com/CTSRD-CHERI/llvm-project/pull/732 into this code and addressed comments from @dpgao and @bsdjhb. This code now depends on changes from `c18n-nobot` in CheriBSD which I believe are contained in this review: https://github.com/CTSRD-CHERI/cheribsd/pull/2090.

dstolfa commented:
With https://github.com/CTSRD-CHERI/cheribsd/pull/2090 having landed, this should now be good to go.

jrtc27 commented on libunwind/src/Registers.hpp:
```diff
@@ -4544,6 +4607,14 @@ class _LIBUNWIND_HIDDEN Registers_riscv {
   reg_t       getIP() const { return _registers[0]; }
   void        setIP(reg_t value) { _registers[0] = value; }
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(_LIBUNWIND_SANDBOX_OTYPES)
+  reg_t       getUnsealedECSP(uintptr_t sealer) { return getSP(); }
```
This makes no sense. CHERI-RISC-V has no such thing as an executive mode, so it should not have methods involving an executive stack pointer.

dstolfa replied:
Yeah it was shim to make it build on CHERI-RISC-V. It's called `ECSP` because it was originally made for Morello, but would probably work similarly in CHERI-RISC-V (assuming otypes end up in there and we make use of them in libunwind). I'll rename usage of `ECSP` to `TrustedStack` or something along those lines to relay what it's supposed to mean. Alternatively, if we really don't want to make use of otypes I could just gate all calls behind a Morello `ifdef`.

jrtc27 commented on libunwind/src/Registers.hpp:
```diff
@@ -4544,6 +4607,14 @@ class _LIBUNWIND_HIDDEN Registers_riscv {
+#ifdef _LIBUNWIND_SANDBOX_HARDENED
+  void        unsealSP(uintptr_t sealer) {}
+  void        unsealFP(uintptr_t sealer) {}
+  void        unsealCalleeSavedRegisters(uintptr_t sealer) {}
```
One should not be able to enable _LIBUNWIND_SANDBOX_OTYPES / _LIBUNWIND_SANDBOX_HARDENED on CHERI-RISC-V, it doesn't make sense to do when there is no c18n for it.

dstolfa replied:
I'll add a guard as I make the other changes.

jrtc27 commented on libunwind/CMakeLists.txt:
```diff
@@ -50,6 +50,8 @@ option(LIBUNWIND_IS_BAREMETAL "Build libunwind for baremetal targets." OFF)
 option(LIBUNWIND_REMEMBER_HEAP_ALLOC "Use heap instead of the stack for .cfi_remember_state." OFF)
 option(LIBUNWIND_INSTALL_HEADERS "Install the libunwind headers." OFF)
+option(LIBUNWIND_SANDBOX_OTYPES "Use a libunwind implementation that assumes a c18n RTLD using otypes." ON)
+option(LIBUNWIND_SANDBOX_HARDENED "Harden the current sandbox implementation." ON)
```
I would rather not have more knobs than we need. We should have one that is "build libunwind for c18n", whatever that means, and whether that's "supports and requires c18n" or just "supports c18n if in use at run time".

jrtc27 replied:
Also bear in mind libunwind isn't CheriBSD-specific, it gets used on Morello Linux too, which won't have support for any of this, and should remain unchanged.

jrtc27 replied:
Renaming it is the right thing to do if it's about c18n not otypes. Whether it should be LIBUNWIND_C18N_SUPPORT or just LIBUNWIND_C18N depends on what happens if you build with the option enabled and run with a non-c18n rtld.

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -703,6 +703,25 @@ Lnovec:
 #elif defined(__aarch64__)
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+DEFINE_LIBUNWIND_FUNCTION(__rtld_unw_setcontext)
```
Comment its prototype above like below?

jrtc27 replied:
Also, does this really need to be in assembly? rtld's overriding implementation isn't.

dstolfa replied:
rtld's overriding implementation is in C because it calls into `unwind_stack()`, but the non-c18n implementation tail calls into the same assembly code as this (https://github.com/CTSRD-CHERI/cheribsd/blob/dev/libexec/rtld-elf/rtld_c18n.c#L856-L859). It's not great, but it was the simplest way to get the values in the right registers without trying to come up with some kind of ABI between libunwind and rtld. It's not great and should probably be reworked at some point in the future, perhaps with the CHERI-RISC-V implementation, but I probably don't have time to do that for the release.

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -703,6 +703,25 @@ Lnovec:
+  ldp    c2, c3,  [c3, #(-0x210 + 0x20)]
+  mov    csp, c16
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+  and    x30, x30, #~1;
```
```suggestion
  and    x30, x30, #~1
```

dstolfa replied:
ACK, good spot, thanks.

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -729,8 +748,8 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
   ldp    c26,c27, [c0, #0x1a0]
   ldp    c28,c29, [c0, #0x1c0]
   ldr    c30,     [c0, #0x200]  // restore pcc into clr
+  add    c16,c0,  #0x230
```
Don't move it unnecessarily

dstolfa replied:
ACK

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -753,14 +772,13 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
+  b      _rtld_unw_setcontext
 #else
-  ret                           // jump to pcc
+  b      _rtld_unw_setcontext_unsealed
```
If _LIBUNWIND_SANDBOX_OTYPES is off surely we can just act like before rather than needing to make a function call?

dstolfa replied:
I think so. I was just avoiding adding code duplication since the thing we'd basically do is the same thing that's in the function call. Alternatively I could simply `#ifdef _LIBUNWIND_SANDBOX_OTYPES` the new code and then `#else` whatever we had there before to keep the functionality 1:1 when the define is not specified, but it would add some duplication.

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -703,6 +703,25 @@ Lnovec:
 
 #elif defined(__aarch64__)
 
+#if defined(__CHERI_PURE_CAPABILITY__)
```
Should this not be gated by _LIBUNWIND_SANDBOX_OTYPES?

dstolfa replied:
It works fine as-is on a CheriBSD rtld that doesn't support c18n, but if the goal is to ifdef the whole thing out and keep libunwind behaving exactly the way it did before (in the sense that not defining `_LIBUNWIND_SANDBOX_OTYPES` causes _all_ the changes to not be compiled in and we're left with the source of libunwind pre-c18n support 1:1) we could put it behind an `#ifdef _LIBUNWIND_SANDBOX_OTYPES` as it simply won't be used unless that is defined.

jrtc27 replied:
I mean, part of the point is that __rtld_foo isn't going to be the naming convention for other OS's dynamic linkers.

dstolfa replied:
ACK. I'm leaning towards leaving it as-is for now because this will likely change as we redesign the APIs post-release anyway. This bit of the code needs to have a cleaner ABI that doesn't depend on the structure of the unwind context. It's very much tied to both libunwind and CheriBSD's rtld. We just need to make sure to document this in the c18n man page.

jrtc27 commented on libunwind/src/CompartmentInfo.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace libunwind {
+class _LIBUNWIND_HIDDEN CompartmentInfo {
+public:
+  static CompartmentInfo sThisCompartmentInfo;
```
Why should there even be an instance of something that has no information nor use for non-CHERI and/or non-c18n?

jrtc27 commented on libunwind/src/libunwind.cpp:
```diff
@@ -28,9 +28,9 @@
+#include "CompartmentInfo.hpp"
 #include "UnwindCursor.hpp"
 
-
```
Don't

jrtc27 commented on libunwind/include/libunwind.h:
```diff
@@ -678,7 +678,8 @@ enum {
   UNW_ARM64_C31 = 229,
-  UNW_ARM64_CSP = 229
+  UNW_ARM64_CSP = 229,
+  UNW_ARM64_ECSP = 230,
```
230 is PCC per https://github.com/ARM-software/abi-aa/blob/main/aadwarf64-morello/aadwarf64-morello.rst#dwarf-register-names

dstolfa commented on libunwind/src/unwind_cheri.h:
```diff
@@ -0,0 +1,36 @@
+#define __UNWIND_CHERI_H__
+
+#ifdef __CHERI_PURE_CAPABILITY__
+#define	_LIBUNWIND_CHERI_PERM_GLOBAL             (1 << 0) /* 0x00000001 */
```
I'm not a big fan of this header existing in the first place, but perhaps for now just `#if defined(__CHERI_PURE_CAPABILITY__) && defined(_LIBUNWIND_TARGET_AARCH64)` will do?

jrtc27 replied:
Why does it even exist? I can't see any references to these defines.

dstolfa reviewed (commented):
Thanks for the detailed feedback! I'd be interested to hear what you think on some of the bits around refactoring/ABIs/ifdefs brought up to avoid ping-ponging due to misunderstanding on what exactly it should look like. In the meantime I'll start making the obvious changes.

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -54,6 +55,16 @@ class DwarfInstructions {
+  static pint_t restoreRegistersFromSandbox(pint_t csp, A &addressSpace,
+                                            R &newRegisters,
+                                            CompartmentInfo &CI, pint_t sealer);
+  static bool isTrampoline(pint_t ecsp, A &addressSpace, CompartmentInfo &CI,
```
There are other types of trampolines in the world. Name should be specific in it being a compartment transition trampoline.

jrtc27 commented on libunwind/src/libunwind.cpp:
```diff
@@ -217,12 +223,24 @@ _LIBUNWIND_WEAK_ALIAS(__unw_get_proc_info, unw_get_proc_info)
+  LocalAddressSpace &addressSpace = LocalAddressSpace::sThisAddressSpace;
+  pint_t sealer = addressSpace.getUnwindSealer();
+  if (addressSpace.isValidSealer(sealer)) {
+#ifdef _LIBUNWIND_SANDBOX_HARDENED
```
Why not just add it to the existing #if? If this isn't defined we do a bunch of queries only to then do nothing with the result.

jrtc27 reviewed (commented):
I don't understand the hardening model and what specifically it's protecting against. What stops me from unwinding into a caller, copying some of the sealed registers, aborting that unwind, starting a new one that stays within my compartment, injecting that register and getting it back unsealed? I am skeptical that the interface can meaningfully be hardened, and that in trying to do so we're adding a whole bunch of complexity that doesn't really buy any security, just an illusion thereof.

The first commit in this series is also a bit screwed up. There's all kinds of mention of sealers despite not having hardening in it, and at least one _LIBUNWIND_SANDBOX_HARDENED check crept in.

jrtc27 commented on libunwind/src/AddressSpace.hpp:
```diff
@@ -408,6 +414,24 @@ inline uint64_t LocalAddressSpace::getRegister(pint_t addr) {
+extern "C" {
+/// Call into the RTLD to get a sealer capability. This sealer will be used to
+/// seal information in the unwinding context.
+uintptr_t _rtld_unw_getsealer(void);
```
```suggestion
uintptr_t _rtld_unw_getsealer();
```
`(void)` and `()` are the same in C++ (and C23), and if they weren't you should've also used the former for the weak fallback.

jrtc27 commented on libunwind/src/CompartmentInfo.hpp:
```diff
@@ -0,0 +1,36 @@
+#if defined(_LIBUNWIND_TARGET_AARCH64)
+  static const uint32_t kNewSPOffset = 12 * sizeof(void *);
+  static const uint32_t kNextOffset = 14 * sizeof(void *);
+  static const uint32_t kFPOffset = 0;
```
Unused

jrtc27 commented on libunwind/src/CompartmentInfo.hpp:
```diff
@@ -0,0 +1,36 @@
+  static const uint32_t kFPOffset = 0;
+  static const uint32_t kCalleeSavedOffset = 2 * sizeof(void *);
+  static const uint32_t kCalleeSavedCount = 10;
+  static const uint32_t kCalleeSavedSize = sizeof(void *);
```
This seems unnecessary? Of course a capability takes a capability's worth of space? And if we also saved floating-point values they'd have a different size. Just use the size of the register being restored in the loop.

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -72,9 +84,8 @@ class DwarfInstructions {
-          (pint_t)((sint_t)registers.getRegister((int)prolog.cfaRegister) +
-                   prolog.cfaRegisterOffset);
+      result = registers.getRegister((int)prolog.cfaRegister);
+      result = (pint_t)((sint_t)result + prolog.cfaRegisterOffset);
```
Not needed now you're not sealing things

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -246,6 +257,93 @@ bool DwarfInstructions<A, R>::getRA_SIGN_STATE(A &addressSpace, R registers,
+              offset);
+  }
+
+  return offset;
```
AFAICT you only need to return this for a debug print; is that really useful, or can we just drop that?

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -246,6 +257,93 @@ bool DwarfInstructions<A, R>::getRA_SIGN_STATE(A &addressSpace, R registers,
+  // Restore: c19-c28
+  for (i = 0, offset = CI.kCalleeSavedOffset; i < CI.kCalleeSavedCount;
+       ++i, offset += CI.kCalleeSavedSize) {
+    pint_t regValue = addressSpace.getCapability(csp + offset);
```
Use getP everywhere over getCapability; the latter is only needed if you're supporting hybrid

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -246,6 +257,93 @@ bool DwarfInstructions<A, R>::getRA_SIGN_STATE(A &addressSpace, R registers,
+    pint_t ecsp, A &addressSpace, CompartmentInfo &CI, pint_t returnAddress) {
+  // TODO(cheri): Use a cfp-based approach rather than the cookie.
+  ptraddr_t expectedReturnAddress =
+      addressSpace.get64(ecsp + CI.kReturnAddressOffset);
```
get<ptraddr_t>?

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -753,15 +771,10 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
-#else
-  ret                           // jump to pcc
-#endif
+  ldr    c2,      [c0, #0x1f0]
```
Why restore c2 and c3 above if they're just going to be clobbered here and restored again for real later?

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -703,6 +703,24 @@ Lnovec:
 #elif defined(__aarch64__)
 
+//
+// extern "C" void __rtld_unw_setcontext(void *c0, void *c1, void *rcsp, void **buf);
```
What's buf and why is it pointing into the middle of the unwind context?

jrtc27 commented on libunwind/src/UnwindRegistersSave.S:
```diff
@@ -861,12 +870,15 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
-  str    c30,     [c0, #0x1e0]
   mov    c1,csp
-  str    c1,      [c0, #0x1f0]
+  stp    c30, c1, [c0, #0x1e0]
```
Leave it as it was to match the upstream non-Morello code

jrtc27 commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -1294,7 +1294,7 @@ class UnwindCursor : public AbstractUnwindCursor{
 #endif // defined(_LIBUNWIND_SUPPORT_TBTAB_UNWIND)
 
-  A               &_addressSpace;
+  A                &_addressSpace;
```
Leave

dpgao commented on libunwind/src/CompartmentInfo.hpp:
```diff
@@ -0,0 +1,34 @@
+  static const uint32_t kCalleeSavedCount = 10;
+  static const uint32_t kReturnAddressOffset = 15 * sizeof(void *) + 8;
+  static const uint32_t kPCOffset = sizeof(void *);
+  static const uint32_t kTrustedFrameSize = 16 * sizeof(void *);
```
Unused constant
```suggestion
```

dpgao commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -246,6 +255,78 @@ bool DwarfInstructions<A, R>::getRA_SIGN_STATE(A &addressSpace, R registers,
+template <typename A, typename R>
+bool DwarfInstructions<A, R>::isCompartmentTransitionTrampoline(
+    pint_t ecsp, A &addressSpace, CompartmentInfo &CI, pint_t returnAddress) {
+  // TODO(cheri): Use a cfp-based approach rather than the cookie.
```
This comment should probably be deleted? The cookie-based approach works well and I don't think there's really a need to change to use CFP in the future.

dpgao commented on libunwind/src/Registers.hpp:
```diff
@@ -1866,14 +1866,18 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
   uintptr_t  getFP() const          { return _registers.__fp; }
   void       setFP(uintptr_t value) { _registers.__fp = value; }
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(_LIBUNWIND_CHERI_C18N_SUPPORT)
+  void       setTrustedStack(uintptr_t value) { _registers.__ecsp = value; }
+  uintptr_t  getTrustedStack() const { return _registers.__ecsp; }
```
First get then set to match the functions above.

```suggestion
  uintptr_t  getTrustedStack() const { return _registers.__ecsp; }
  void       setTrustedStack(uintptr_t value) { _registers.__ecsp = value; }
```

jrtc27 commented on libunwind/CMakeLists.txt:
```diff
@@ -293,6 +295,19 @@ if (NOT LIBUNWIND_ENABLE_THREADS)
 
+# Sandboxing and c18n support
+if (LIBUNWIND_SANDBOX_OTYPES)
+  if (NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64c" OR NOT CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
```
Please fix this so cheribuild can enable c18n support when building libunwind standalone (you can test it with `cheribuild libunwind-morello-purecap --libunwind-morello-purecap/cmake-options=-DLIBUNWIND_CHERI_C18N_SUPPORT`)

jrtc27 commented on libunwind/include/libunwind.h:
```diff
@@ -678,7 +678,11 @@ enum {
-  UNW_ARM64_CSP = 229
+  UNW_ARM64_CSP = 229,
+  // Use 240 for ECSP (executive stack pointer). ECSP is not a real DWARF
+  // register, but we need it to implemented c18n-aware unwinding. We pick 240
```
```suggestion
  // register, but we need it to implement c18n-aware unwinding. We pick 240
```

jrtc27 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -246,6 +255,77 @@ bool DwarfInstructions<A, R>::getRA_SIGN_STATE(A &addressSpace, R registers,
+  CHERI_DBG("SANDBOX: SETTING RESTRICTED CSP: %#p\n",
+            (void *)newRegisters.getSP());
+  // Restore callee-saved registers
+  size_t i;
+  size_t offset;
+  // Restore: c19-c28
+  for (i = 0, offset = CI.kCalleeSavedOffset; i < CI.kCalleeSavedCount;
```
```suggestion
  // Restore: c19-c28
  for (size_t i = 0, offset = CI.kCalleeSavedOffset; i < CI.kCalleeSavedCount;
```

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -713,7 +732,8 @@ Lnovec:
   // skip restore of c0,c1 for now
-  ldp    c2, c3,  [c0, #0x020]
+  // also skip restoring c2 and c3 because they will get clobbered later on
+  // when supporting c18n-aware unwinding
```
Not just when, they always do, otherwise it wouldn't work

jrtc27 reviewed (commented):
Just some nits and a build system issue left

jrtc27 commented on libunwind/src/UnwindRegistersSave.S:
```diff
@@ -885,7 +898,7 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
   str    d31,     [c0, #0x0f8]
   mov    x0, #0                   // return UNW_ESUCCESS
-  ret
+  b      _rtld_unw_getcontext
```
Thinking about this more, won't PLTs and lazy binding screw this up? c16 and c17 can be clobbered by PLTs, and lazy binding can clobber more.

dstolfa replied:
I'm not sure I follow. `_rtld_unw_getcontext` doesn't use c16, and `_rtld_unw_setcontext` only uses it inside the function itself to temporarily store the value of c2. Why does it matter if the PLT makes use of c16 and c17?

In the `unw_getcontext` path we `stp c16,c17, [c0, #0x100]` first, then we branch to `_rtld_unw_getcontext`, going through the PLT (which clobbers c16 and c17, but we have the old values at c0+0x100), but gives us a sealed executive stack pointer.

In `_rtld_unw_setcontext`, c16 and c17 don't get restored anyway since they were clobbered (there's already a comment stating that there), then we go through the PLT and clobber them again, and then in the assembly inside rtld we clobber c16 yet again to temporarily store the value of c2.

Perhaps I'm missing something here? If you mean that the call to `unw_getcontext` would go via the PLT, that is handled by marking `unw_getcontext` as a trusted function in the c18n policy, so it should never go through the PLT.

jrtc27 replied:
Oh I meant to comment on the setcontext one, just distracted by the concurrent meeting. In that case IP0/IP1 aren't a concern, but lazy binding still is. Specifically, temporary registers aren't saved by _rtld_bind_start, because they're not meant to be live across calls. To avoid that you'd either need to make an indirect call or backport and use variant PCS support (https://reviews.freebsd.org/D44869 has the stack).

jrtc27 replied:
(Also, re "marking unw_getcontext as a trusted function in the c18n policy, so it should never go through the PLT", it'll still go though a PLT, just not through an additional compartment switch trampoline)

dstolfa replied:
> Oh I meant to comment on the setcontext one, just distracted by the concurrent meeting. In that case IP0/IP1 aren't a concern, but lazy binding still is. Specifically, temporary registers aren't saved by _rtld_bind_start, because they're not meant to be live across calls. To avoid that you'd either need to make an indirect call or backport and use variant PCS support (https://reviews.freebsd.org/D44869 has the stack).

Ah. That's slightly annoying. How do we want to approach this for the release? Both of these sound a bit risky w.r.t. regressions. It seems unlikely that exception handling code will care about this in any meaningful way, but it does break libunwind API semantics...

jrtc27 commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -753,14 +772,17 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
+  ldp    c0, c1,  [c0, #0x000]
+  // XXX: variant PCS is not yet supported by rtld, work around it
+  // using a function pointer.
+  adrp c16, :got:_rtld_unw_setcontext
```
Align the operands like above for these instructions

jrtc27 reviewed (approved):
One style nit, otherwise this is fine for "v0"

jrtc27 commented:
Feel free to merge once both Linux and FreeBSD Jenkins jobs are done (though something's gone very wrong if that change has broken something...)
